lib/rspec/abq/extensions.rb

```diff
@@ -110,10 +110,13 @@ module Runner
         # @return [Fixnum] exit status code. 0 if all specs passed,
         #   or the configured failure exit code (1 by default) if specs
         #   failed.
-        def run_specs(example_groups)
+        def run_specs(_example_groups)
           should_quit = RSpec::Abq.setup_after_specs_loaded!
           return 0 if should_quit
 
+          # rspec-abq pulls the ordering from the init-message. Here we ensure the example groups are in the same ordering.
+          example_groups = @world.ordered_example_groups
```

Comment:
Are we sure that this is the same as the `example_groups` passed in to `run_specs`? That is, no other filtering on the world example groups is done before calling run_specs?

Reply:
answering my own question: this appears to be safe

https://github.com/rspec/rspec-core/blob/522b7727d02d9648c090b56fa68bbdc18a21c04d/lib/rspec/core/runner.rb#L85-L92

will add a comment

Reply:
@michaelglass can you explain this change in more detail? If this line is the same as what is passed to `run_specs`, what'd we actually change/fix with this PR?

Reply:
was answered in slack https://rwxhq.slack.com/archives/C04BKCZHM4H/p1670904034332249

The difference between this and the argument is that we're potentially re-shuffling (or even un-shuffling, although unlikely) the example groups AFTER abq-initialization has registered an ordering it fetched from the manifest.

The regular use case here is this invocation:
```
abq test -- bundle exec rspec --order rand
```

in this case, the manifest includes a random seed, and the workers use the manifest's seed to ensure consistent test ordering (rather than using the random seed that the rspec invocation provides).

It's possible, although unlikely that workers may have different rspec configuration files. Always reshuffling after setting up the order from the manifest ensures that no matter how rspec was started, each rspec will have the ordering specified in the manifest.

## alternative strategies:

### only reshuffle if we detect that we've changed the ordering
this is an easy performance optimization

### initialize rspec-abq higher in the call chain
We [tried this and couldn't find a solution](https://github.com/rwx-research/rspec-abq/pull/14/) without having the user also explicitly requiring `rspec/abq` . Maybe worth a further investigation.

#### change `rspec-abq`'s strategy to seek-to-test rather than run-in-order (e.g. ignore ordering in tests)
This is an interesting approach, and would unlock a bunch of interesting strategies from the abq side (e.g. running flakey tests twice), but it would be a lot of work and has a bunch of rough edges around test setup & teardown.

Reply:
https://github.com/rwx-research/rspec-abq/pull/66

Reply:
As far as I can tell, there are two ways that rspec would normally requires abq:

1. via a `--require` directive in the CLI / in the `.rspec` file.
     in this case, RSpec::Abq is available [after options are handled here](https://github.com/rspec/rspec-core/blob/main/lib/rspec/core/configuration_options.rb#L22)
2. via transitive requires through the `spec_helper.rb`.
     in this case, RSpec::Abq is available [after specs are loaded, here](https://github.com/rspec/rspec-core/blob/main/lib/rspec/core/runner.rb#LL102C24-L102C32)
  
  ... ok I have one idea. let's see if it works

Reply:
Are the places you references in code where the order gets set though? RSpec supports this:

```ruby
RSpec.configure do |rspec|
  rspec.register_ordering(:global) do |items|
    items.reverse
  end
end
```

Which would presumably go into a `spec_helper.rb` and affect how the specs are ordered. I feel like the actual ordering must happen closer to when the tests are run, not when rspec is loaded, otherwise I'm not sure how defining ordering in an `RSpec.configure` block could work.

I'm pretty sure ordering happens here

https://github.com/rspec/rspec-core/blob/522b7727d02d9648c090b56fa68bbdc18a21c04d/lib/rspec/core/runner.rb#L89

Which means that we can hook into `run` to override the configuration before the original `run` calls `run_specs`

Reply:
Looking closer at rspec-abq's implementation, I'm almost certain the issue is that we're doing all of the ABQ initialization after `run_specs` is called, instead of doing it in `run` before `run_specs` is called. I strongly think we should initialize ABQ at a higher level so that `run_specs` can be called as it normally is, to prevent other unexpected issues when RSpec incompatibility with ABQ's modifications.

Reply:
Plus, in our current implementation, ordering is happening twice: once in `run` and then again in our overridden `run_specs`. If we move our hooks, then ordering will only happen once.

Reply:
Hmm.  I think you have something here. It'll require us splitting off our manifest logic from the rest of the initialization logic.

Reply:
Yeah the problem was conflating two concepts:
- initialization that requires all specs being loaded (specifically : manifest generation)
- initialization that does not require specs being loaded (everything else)

we have to run manifest generation somewhere explicitly after specs are loaded but before specs are run. The earliest place possible is `World#ordered_example`. I thought `#setup` would be good too, but unfortunately `rspec-abq` is not loaded before `#setup` is run (in the latest, case, it's only loaded after specs are run which happens in the last line of `#setup`)

Incidentally, we were running _the rest of setup, including, for instance, ordering, at the same place we were writing the manifest.

In fact, we can run the rest of setup as soon as `setup-abq` is loaded, as long as the manifest isn't being written.

anyways thanks for the push back! I think that's the solution.

Reply:
https://github.com/rwx-research/rspec-abq/pull/70